core/src/main/java/fr/sncf/osrd/conflicts/IncrementalConflictDetector.kt

```diff
@@ -0,0 +1,150 @@
+import fr.sncf.osrd.standalone_sim.result.ResultTrain.SpacingRequirement
+import java.util.TreeMap
+import kotlin.math.max
+import kotlin.math.min
+
+const val DEFAULT_WORK_SCHEDULE_ID: Long = -1
+
+sealed interface IncrementalConflictResponse
+
+data class ConflictResponse(
+    // minimum delay that should be added to the train so that there are no conflicts anymore
+    val minDelayWithoutConflicts: Double,
+    // Time at which the first conflict happened
+    val firstConflictTime: Double,
+) : IncrementalConflictResponse
+
+data class NoConflictResponse(
+    // maximum delay that can be added to the train without creating any conflict
+    val maxDelayWithoutConflicts: Double,
+    // minimum begin time of the next requirement that could conflict
+    val timeOfNextConflict: Double,
+) : IncrementalConflictResponse
+
+fun incrementalConflictDetectorFromTrainReq(
+    requirements: List<TrainRequirements>
+): IncrementalConflictDetector {
+    return IncrementalConflictDetector(convertTrainRequirements(requirements))
+}
+
+fun incrementalConflictDetectorFromReq(
+    requirements: List<Requirements>
+): IncrementalConflictDetector {
+    return IncrementalConflictDetector(requirements)
+}
+
+/**
+ * This class takes a list of requirements as input, and can only be used to compare them to a *new*
+ * set of requirements. The initial requirements cannot be modified. Conflicts between initial
+ * trains are not tested.
+ *
+ * In practice, this is used for STDCM, where the initial requirements represent the timetable
+ * trains, and new requirements come from the train we are trying to fit in the timetable.
+ */
+class IncrementalConflictDetector(requirements: List<Requirements>) {
+    // Zone name -> (end time -> Range(start time, end time))
+    private val spacingZoneUses: Map<String, TreeMap<Double, Range<Double>>>
+
+    init {
+        spacingZoneUses = generateSpacingZoneUses(requirements)
+    }
+
+    private fun generateSpacingZoneUses(
+        requirements: List<Requirements>
+    ): Map<String, TreeMap<Double, Range<Double>>> {
+        // We first create RangeSets to handle the overlaps, but then
+        // convert them to TreeMaps (`.higherEntry` is extremely convenient here)
+        val rangeSets = mutableMapOf<String, RangeSet<Double>>()
+        for (req in requirements) {
+            for (spacingReq in req.spacingRequirements) {
+                val set = rangeSets.computeIfAbsent(spacingReq.zone) { TreeRangeSet.create() }
+                set.add(Range.closedOpen(spacingReq.beginTime, spacingReq.endTime))
+            }
+        }
+        return rangeSets
+            .map { it.key to TreeMap(it.value.asRanges().associate { it.upperEndpoint() to it }) }
+            .toMap()
+    }
+
+    /**
+     * Check for any conflicts between the initial requirements and the ones given here as method
+     * input. Returns a polymorphic response with different extra data for either case (conflict /
+     * no conflict).
```

Comment:
nit (Checks/Returns consistency + any):
```suggestion
     * Checks for any conflict between the initial requirements and the ones given here as method
     * input. Returns a polymorphic response with different extra data for either case (conflict /
     * no conflict).
```